GetMoarFediverse Actions fail on network connection issues

The DDoS against mastodon.social and mastodon.online is causing GetMoarFediverse to fail on network connection errors. I notice this as failed GitHub Actions. It would be ideal if the action would keep running even if an instance is misbehaving.

After looking through the Actions logs for several failed runs, I found a few variations, but they all seem to be related to network connectivity. This one mentions mastodon.online specifically, and happened during the DDoS:

```Run g3rv4/GetMoarFediverse@main
Run echo "MOAR_TMP_DIR=$(mktemp -d)" >> $GITHUB_ENV
Run cp /home/runner/work/GetMoarFediverse-Action/GetMoarFediverse-Action/config.json ${MOAR_TMP_DIR}/config.json
Run dawidd6/action-download-artifact@v2
==> Repository: 33b5e5/GetMoarFediverse-Action
==> Artifact name: moar-imported
==> Local path: /tmp/tmp.VFEAu76DBN
==> Workflow name: 44021760
==> Workflow conclusion: success
==> (found) Run ID: 4050623493
==> (found) Run date: 2023-01-31T04:02:03Z
==> Artifact: 534127211
==> Downloading: moar-imported.zip (336.1 kB)
==> Extracting: moar-imported.zip
Run wget -nv https://github.com/g3rv4/GetMoarFediverse/releases/latest/download/GetMoarFediverse_linux-x64.tgz
2023-01-31 04:46:34 URL:https://objects.githubusercontent.com/github-production-release-asset-2e65be/573569752/f55f61f2-f98d-4e39-bcf4-85b8c1941d5d?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230131%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230131T044634Z&X-Amz-Expires=300&X-Amz-Signature=b9d8e80bbe00ce278769439ae27545978e7bd683679ec55490735a471572caa6&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=573569752&response-content-disposition=attachment%3B%20filename%3DGetMoarFediverse_linux-x64.tgz&response-content-type=application%2Foctet-stream [8312364/8312364] -> "GetMoarFediverse_linux-x64.tgz" [1]
Unhandled exception. System.Net.Http.HttpRequestException: Connection reset by peer (mastodon.online:80)
 ---> System.Net.Sockets.SocketException (104): Connection reset by peer
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError , CancellationToken )
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 )
   at System.Net.Sockets.Socket.<ConnectAsync>g__WaitForConnectWithCancellation|277_0(AwaitableSocketAsyncEventArgs , ValueTask , CancellationToken )
   at System.Net.Http.HttpConnectionPool.ConnectToTcpHostAsync(String , Int32 , HttpRequestMessage , Boolean , CancellationToken )
   --- End of inner exception stack trace ---
   at System.Net.Http.HttpConnectionPool.ConnectToTcpHostAsync(String , Int32 , HttpRequestMessage , Boolean , CancellationToken )
   at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage , Boolean , CancellationToken )
   at System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage , Boolean , CancellationToken )
   at System.Net.Http.HttpConnectionPool.AddHttp11ConnectionAsync(HttpRequestMessage )
   at System.Threading.Tasks.TaskCompletionSourceWithCancellation`1.WaitWithCancellationAsync(CancellationToken )
   at System.Net.Http.HttpConnectionPool.GetHttp11ConnectionAsync(HttpRequestMessage , Boolean , CancellationToken )
   at System.Net.Http.HttpConnectionPool.SendWithVersionDetectionAndRetryAsync(HttpRequestMessage , Boolean , Boolean , CancellationToken )
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage , Boolean , CancellationToken )
   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage , HttpCompletionOption , CancellationTokenSource , Boolean , CancellationTokenSource , CancellationToken )
   at TurnerSoftware.RobotsExclusionTools.RobotsFileParser.FromUriAsync(Uri robotsUri, RobotsFileAccessRules accessRules, CancellationToken cancellationToken) in /home/runner/work/RobotsExclusionTools/RobotsExclusionTools/src/TurnerSoftware.RobotsExclusionTools/RobotsFileParser.cs:line 96
   at Program.<>c__DisplayClass0_0.<<<Main>$>b__0>d.MoveNext() in /var/src/Program.cs:line 44
--- End of stack trace from previous location ---
   at System.Threading.Tasks.Parallel.<>c__50`1.<<ForEachAsync>b__50_0>d.MoveNext()
--- End of stack trace from previous location ---
   at Program.<Main>$(String[] args) in /var/src/Program.cs:line 40
   at Program.<Main>(String[] args)
/home/runner/work/_temp/b6f48dec-8456-4a54-9c40-1d20e9d0b9c7.sh: line 4:  1727 Aborted                 (core dumped) ./GetMoarFediverse /tmp/tmp.VFEAu76DBN/config.json
Error: Process completed with exit code 134.
```

This one mentions `RobotsFileParser.cs` so I guess it could be related to robots.txt:

```Run g3rv4/GetMoarFediverse@main
Run echo "MOAR_TMP_DIR=$(mktemp -d)" >> $GITHUB_ENV
Run cp /home/runner/work/GetMoarFediverse-Action/GetMoarFediverse-Action/config.json ${MOAR_TMP_DIR}/config.json
Run dawidd6/action-download-artifact@v2
==> Repository: 33b5e5/GetMoarFediverse-Action
==> Artifact name: moar-imported
==> Local path: /tmp/tmp.SlmAtOz1Nn
==> Workflow name: 44021760
==> Workflow conclusion: success
==> (found) Run ID: 4050623493
==> (found) Run date: 2023-01-31T04:02:03Z
==> Artifact: 534127211
==> Downloading: moar-imported.zip (336.1 kB)
==> Extracting: moar-imported.zip
Run wget -nv https://github.com/g3rv4/GetMoarFediverse/releases/latest/download/GetMoarFediverse_linux-x64.tgz
2023-01-31 05:31:44 URL:https://objects.githubusercontent.com/github-production-release-asset-2e65be/573569752/f55f61f2-f98d-4e39-bcf4-85b8c1941d5d?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230131%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230131T053143Z&X-Amz-Expires=300&X-Amz-Signature=c961beec16d7b82bd6fc5ba071332b22b79949e1875d14f3c24bd617a8d95ad6&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=573569752&response-content-disposition=attachment%3B%20filename%3DGetMoarFediverse_linux-x64.tgz&response-content-type=application%2Foctet-stream [8312364/8312364] -> "GetMoarFediverse_linux-x64.tgz" [1]
Unhandled exception. System.Net.Http.HttpRequestException: An error occurred while sending the request.
 ---> System.IO.IOException: Unable to read data from the transport connection: Connection reset by peer.
 ---> System.Net.Sockets.SocketException (104): Connection reset by peer
   --- End of inner exception stack trace ---
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError , CancellationToken )
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource<System.Int32>.GetResult(Int16 )
   at System.Net.Http.HttpConnection.InitialFillAsync(Boolean )
   at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage , Boolean , CancellationToken )
   --- End of inner exception stack trace ---
   at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage , Boolean , CancellationToken )
   at System.Net.Http.HttpConnectionPool.SendWithVersionDetectionAndRetryAsync(HttpRequestMessage , Boolean , Boolean , CancellationToken )
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage , Boolean , CancellationToken )
   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage , HttpCompletionOption , CancellationTokenSource , Boolean , CancellationTokenSource , CancellationToken )
   at TurnerSoftware.RobotsExclusionTools.RobotsFileParser.FromUriAsync(Uri robotsUri, RobotsFileAccessRules accessRules, CancellationToken cancellationToken) in /home/runner/work/RobotsExclusionTools/RobotsExclusionTools/src/TurnerSoftware.RobotsExclusionTools/RobotsFileParser.cs:line 96
   at Program.<>c__DisplayClass0_0.<<<Main>$>b__0>d.MoveNext() in /var/src/Program.cs:line 44
--- End of stack trace from previous location ---
   at System.Threading.Tasks.Parallel.<>c__50`1.<<ForEachAsync>b__50_0>d.MoveNext()
--- End of stack trace from previous location ---
   at Program.<Main>$(String[] args) in /var/src/Program.cs:line 40
   at Program.<Main>(String[] args)
/home/runner/work/_temp/0d4f1533-db12-43d7-954f-120dccc8d671.sh: line 4:  1724 Aborted                 (core dumped) ./GetMoarFediverse /tmp/tmp.SlmAtOz1Nn/config.json
Error: Process completed with exit code 134.
```

Here's a failure after what looks like an SSL validation issue:

```
Run g3rv4/GetMoarFediverse@main
Run echo "MOAR_TMP_DIR=$(mktemp -d)" >> $GITHUB_ENV
Run cp /home/runner/work/GetMoarFediverse-Action/GetMoarFediverse-Action/config.json ${MOAR_TMP_DIR}/config.json
Run dawidd6/action-download-artifact@v2
==> Repository: 33b5e5/GetMoarFediverse-Action
==> Artifact name: moar-imported
==> Local path: /tmp/tmp.IMLPS2uFmn
==> Workflow name: 44021760
==> Workflow conclusion: success
==> (found) Run ID: 4093268788
==> (found) Run date: 2023-02-04T20:20:01Z
==> Artifact: 541832030
==> Downloading: moar-imported.zip (336.16 kB)
==> Extracting: moar-imported.zip
Run wget -nv https://github.com/g3rv4/GetMoarFediverse/releases/latest/download/GetMoarFediverse_linux-x64.tgz
2023-02-04 20:31:39 URL:https://objects.githubusercontent.com/github-production-release-asset-2e65be/573569752/f55f61f2-f98d-4e39-bcf4-85b8c1941d5d?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230204%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230204T203139Z&X-Amz-Expires=300&X-Amz-Signature=db5591cde5bdd7355af8750974b3d7c130c8449773a493a9a6cbd566c362e61d&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=573569752&response-content-disposition=attachment%3B%20filename%3DGetMoarFediverse_linux-x64.tgz&response-content-type=application%2Foctet-stream [8312364/8312364] -> "GetMoarFediverse_linux-x64.tgz" [1]
Unhandled exception. System.Net.Http.HttpRequestException: The SSL connection could not be established, see inner exception.
 ---> System.Security.Authentication.AuthenticationException: The remote certificate is invalid because of errors in the certificate chain: NotTimeValid
   at System.Net.Security.SslStream.SendAuthResetSignal(ProtocolToken , ExceptionDispatchInfo )
   at System.Net.Security.SslStream.CompleteHandshake(SslAuthenticationOptions )
   at System.Net.Security.SslStream.ForceAuthenticationAsync[TIOAdapter](TIOAdapter , Boolean , Byte[] , Boolean )
   at System.Net.Http.ConnectHelper.EstablishSslConnectionAsync(SslClientAuthenticationOptions , HttpRequestMessage , Boolean , Stream , CancellationToken )
   --- End of inner exception stack trace ---
   at System.Net.Http.ConnectHelper.EstablishSslConnectionAsync(SslClientAuthenticationOptions , HttpRequestMessage , Boolean , Stream , CancellationToken )
   at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage , Boolean , CancellationToken )
   at System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage , Boolean , CancellationToken )
   at System.Net.Http.HttpConnectionPool.AddHttp11ConnectionAsync(HttpRequestMessage )
   at System.Threading.Tasks.TaskCompletionSourceWithCancellation`1.WaitWithCancellationAsync(CancellationToken )
   at System.Net.Http.HttpConnectionPool.GetHttp11ConnectionAsync(HttpRequestMessage , Boolean , CancellationToken )
   at System.Net.Http.HttpConnectionPool.SendWithVersionDetectionAndRetryAsync(HttpRequestMessage , Boolean , Boolean , CancellationToken )
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage , Boolean , CancellationToken )
   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage , HttpCompletionOption , CancellationTokenSource , Boolean , CancellationTokenSource , CancellationToken )
   at TurnerSoftware.RobotsExclusionTools.RobotsFileParser.FromUriAsync(Uri robotsUri, RobotsFileAccessRules accessRules, CancellationToken cancellationToken) in /home/runner/work/RobotsExclusionTools/RobotsExclusionTools/src/TurnerSoftware.RobotsExclusionTools/RobotsFileParser.cs:line 96
   at Program.<>c__DisplayClass0_0.<<<Main>$>b__0>d.MoveNext() in /var/src/Program.cs:line 44
--- End of stack trace from previous location ---
   at System.Threading.Tasks.Parallel.<>c__50`1.<<ForEachAsync>b__50_0>d.MoveNext()
--- End of stack trace from previous location ---
   at Program.<Main>$(String[] args) in /var/src/Program.cs:line 40
   at Program.<Main>(String[] args)
/home/runner/work/_temp/901b9cde-e752-43ed-b847-3616921527b7.sh: line 4:  1666 Aborted                 (core dumped) ./GetMoarFediverse /tmp/tmp.IMLPS2uFmn/config.json
Error: Process completed with exit code 134.
```

--- a/src/Program.cs
+++ b/src/Program.cs
@@ -41,7 +41,14 @@ await Parallel.ForEachAsync(Config.Instance.Sites,
     new ParallelOptions { MaxDegreeOfParallelism = Config.Instance.Sites.Length },
     async (site, _) =>
     {
-        sitesRobotFile[site.Host] = await robotsFileParser.FromUriAsync(new Uri($"http://{site.Host}/robots.txt"));
+        try
+        {
+            sitesRobotFile[site.Host] = await robotsFileParser.FromUriAsync(new Uri($"http://{site.Host}/robots.txt"));
+        }
+        catch
+        {
+            Console.WriteLine($"Ignoring {site.Host} because had issues fetching its robots data (is the site down?)");
+        }
     }
 );
 
@@ -87,6 +94,11 @@ await Parallel.ForEachAsync(sitesTags, new ParallelOptions{MaxDegreeOfParallelis
             Console.WriteLine($"Scraping {url} is not allowed based on their robots.txt file");
             return;
         }
+    }
+    else
+    {
+        Console.WriteLine($"Not scraping {url} because I couldn't fetch robots data.");
+        return;
     }
     
     HttpResponseMessage? response = null;

